Add getTestLogger injection seam for tests

markfields opened the pull request:
For now, only updating one test file in end-to-end tests, and also the stress test, to prove it out. Subsequent PRs will include `getTestLogger` throughout end-to-end tests.

markfields committed:
Simplify env vars for identifying test logger pkg

markfields committed:
Add first usage of getTestLogger

markfields committed:
lock

markfields committed:
REVERT - prove that injection works

markfields committed:
Revert "REVERT - prove that injection works"

This reverts commit 166b3926f95eb62e111f7c2c4896e129dedfe163.

markfields commented on packages/utils/telemetry-utils/src/index.ts:
```diff
@@ -2,6 +2,15 @@
+
+import { ITelemetryBaseLogger } from "@fluidframework/common-definitions";
+
+//* Where should this actually go? common-definitions?
+declare global {
+    /** This function may be provided by the environment, like a mocha test hook */
+    export const getTestLogger: (() => ITelemetryBaseLogger) | undefined;
+}
+
```
@anthony-murphy - Thoughts on where this declaration should go? common-definitions alongside `ITelemetryBaseLogger`? Or some other package?  Would it make sense to consolidate this and `getFluidTestDriver` in a package called something like `@fluid-test/test-seams`?

anthony-murphy replied:
i'm leaning towards test-driver-definitions. i actually debated just calling it test definitions, which may have been the right approach

markfields replied:
K makes sense to me, I'll rename the package.

markfields replied:
...in a future PR

anthony-murphy commented on packages/utils/telemetry-utils/src/index.ts:
```diff
@@ -2,6 +2,15 @@
+//* Where should this actually go? common-definitions?
+declare global {
+    /** This function may be provided by the environment, like a mocha test hook */
+    export const getTestLogger: (() => ITelemetryBaseLogger) | undefined;
```
i'd prefer it this couldn't be undefined, basically it needs to exist. you could actually create a dummy in mocha-test-setup, which already overrides console and such, so i think it fits

markfields replied:
Nice, I wanted that too but didn't spend much time thinking about where/how.

markfields committed:
declare getTestLogger as always defined and in different pkg

markfields committed:
Inject logger to nodeStressTest, and add flush to interface

markfields committed:
Commit to synchronous flush

markfields committed:
Fix (hopefully) mocha hooks

And add singleton parameter to IBufferedTelemetryLogger

markfields committed:
"Fix" flush mechanics

With a setTimeout promise. Will revisit in a future change

anthony-murphy commented on packages/test/test-driver-definitions/src/index.ts:
```diff
@@ -3,13 +3,14 @@
     export function getFluidTestDriver(): ITestDriver;
+
+    /** This function may be provided by the environment, like a mocha test hook */
+    export function getTestLogger(singleton?: boolean): ITelemetryBufferedLogger;
```
why do we have to choose if a singleton or not?

anthony-murphy replied:
looks to me like this can always be a singleton, and we can get rid of the parameter

markfields replied:
Hmm that's true, thanks!

markfields committed:
Switch flush back to async

Turns out we need to give the upload request a chance to clear,
so this API needs to be async after all

markfields committed:
Remove singleton parameter

markfields committed:
Merge remote-tracking branch 'origin/main' into logger-injection

markfields committed:
Nits and fixes

anthony-murphy commented on packages/test/mocha-test-setup/src/mochaHooks.ts:
```diff
@@ -16,4 +32,9 @@ export const mochaHooks = {
     },
+    async afterAll() {
+        // Flush the logs before exiting
+        const logger: ITelemetryBufferedLogger = getTestLogger();
```
don;t think we need this. the null logger will never need to flush, and the thing inserting a real logger should do this instead

markfields committed:
Fix mocha-test-setup hooks

markfields committed:
lock

markfields committed:
move back to beforeall

markfields commented:
> Baseline CI build failed, cannot generate bundle analysis at this time
> 
> Baseline commit: [1f8a49e](https://github.com/microsoft/FluidFramework/commit/1f8a49e9fa2ee91312f2cabdb4b6b255c094d763)
> 
> Generated by 🚫 [dangerJS](https://danger.systems/js) against [64de1d1](https://github.com/microsoft/FluidFramework/commit/64de1d1c0f18c4ea7c964127fec41897aa283052)

@heliocliu is this concerning?

markfields committed:
PR nits

heliocliu commented:
> 
> 
> > Baseline CI build failed, cannot generate bundle analysis at this time
> > Baseline commit: [1f8a49e](https://github.com/microsoft/FluidFramework/commit/1f8a49e9fa2ee91312f2cabdb4b6b255c094d763)
> > Generated by 🚫 [dangerJS](https://danger.systems/js) against [64de1d1](https://github.com/microsoft/FluidFramework/commit/64de1d1c0f18c4ea7c964127fec41897aa283052)
> 
> @heliocliu is this concerning?

Nope, error says the build for the baseline commit in the pipe failed so it doesn't have anything to compare.  Pipeline logs look like that failed because ADO services were down.  I'll kick off the pipe again in case anyone else hits this

anthony-murphy commented on packages/test/end-to-end-tests/src/test/container.spec.ts:
```diff
@@ -18,20 +18,22 @@ import { MockDocumentDeltaConnection } from "@fluid-internal/test-loader-utils";
 import { ensureFluidResolvedUrl } from "@fluidframework/driver-utils";
 import { requestFluidObject } from "@fluidframework/runtime-utils";
-import { ITestDriver } from "@fluidframework/test-driver-definitions";
+import { ITestDriver, ITelemetryBufferedLogger } from "@fluidframework/test-driver-definitions";
```
do we actually need to expose ITelemetryBufferedLogger, as the mocha hook should flush, keeping it transparent for the test

markfields replied:
Yeah I debated about that too... We can discuss and do a follow-up to get rid of that interface on this side if we want.

markfields replied:
Oh right - we have to leave the interface on the global declare so that the stress test can flush.  If we want to adopt a convention of giving `logger` in mocha files the smaller type, I'm fine with that.  Again, in a future change.  Going to merge this as soon as my test pipeline run finishes.

markfields committed:
Fix a few bugs

markfields committed:
Merge remote-tracking branch 'origin/main' into logger-injection
